### PR TITLE
Sort to-dos

### PR DESCRIPTION
**Pull request by mcgBradley**

### CODE REVIEW

File: app/controllers/todos_controller.rb [hunk 1]
```diff
@@ -1,7 +1,21 @@
 class TodosController < ApplicationController
+    @@sort_Type = "none"
     before_action :authenticate
     def index
-        @todos = current_user.todos
+        if params[:sort]
```
Comment (v1dya) [inline app/controllers/todos_controller.rb]: Why is this if statement here?

File: app/controllers/todos_controller.rb [hunk 2]
```diff
@@ -1,7 +1,21 @@
 class TodosController < ApplicationController
+    @@sort_Type = "none"
```
Comment (v1dya) [inline app/controllers/todos_controller.rb]: Are we sure we want a class variable? Why not an instance variable? How would it work with multiple users?

File: app/controllers/todos_controller.rb [hunk 3]
```diff
@@ -1,7 +1,21 @@
...
     before_action :authenticate
     def index
-        @todos = current_user.todos
+        if params[:sort]
+            @@sort_Type = params[:sort]
+        end
+        if @@sort_Type == "none"
```
Comment (v1dya) [inline app/controllers/todos_controller.rb]: Can we use a more efficient way of finding sort modes than `if` and `elsif` statements?

File: spec/features/user_sorts_todos_by_alphabetical_spec.rb
Comment (v1dya) [inline spec/features/user_sorts_todos_by_alphabetical_spec.rb:1]: If different specs are testing the same select button in your app then the better convention is to use a single file with many different `feature` blocks or `context` blocks. 

A lot of these specs have some repeated code which you can optimise with DRY.

Your tests should also test what happens in edge cases. What happens if there are 0 todos? What happens if a todo starts with a number and you sort alphabetically? What happens if there is a todo with an empty string like ''? Think of as many edge cases as you can and write tests for them.

For Rspec testing convention and best practices most of us use https://www.betterspecs.org/. It's not an absolute rule that you have to follow every convention but most times it's good to follow. In some rare cases it will make sense to not follow convention but you should have a good reason for it

File: spec/features/user_visits_homepage_spec.rb [hunk 4]
```diff
@@ -4,6 +4,6 @@
     scenario "successfully" do
         visit root_path
 
-        expect(page).to have_css('h1', text: 'Todos')
+        expect(page).to have_css('h1', text: 'To-dos')
```
Comment (v1dya) [inline spec/features/user_visits_homepage_spec.rb]: This generally would not be part of this PR. A PR should be for one feature or some related features. It's good that you found this change but you should make a separate PR to merge this in

File: spec/support/features/select_matcher.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+RSpec::Matchers.define :appear_before do |later_content|
```
Comment (v1dya) [inline spec/support/features/select_matcher.rb:1]: Great work! This is a nice use of matchers

File: app/views/todos/index.html.erb [hunk 6]
```diff
@@ -10,4 +35,18 @@
...
+</ul>
+
+<script>
+
+    document.getElementById('sort').onchange = function () {
+        console.log(this.value);
+        localStorage.setItem('sort_type', this.value);
+    }
+
+    window.onload = function () {
```
Comment (v1dya) [inline app/views/todos/index.html.erb:47]: Great solution for state management. This will be easier in react so when you do get some tickets on our frontend you won't have to struggle this much

File: app/views/todos/index.html.erb
Comment (v1dya) [inline app/views/todos/index.html.erb:1]: Could we add some CSS to make it look a bit nicer? Maybe center it and add some colours for buttons or something? Anything works make it look fun

**Review by v1dya — CHANGES_REQUESTED** [review]
Sorry about the 1000 emails I forgot the single comment feature sends it immediately on Github.

Great work on your first PR! A lot of your basics are really good and the logic you used is great, all you need to learn is the rails way and how to efficiently use it. Most of my comments are convention related which you will get used to.

looking forward to the next revision!

File: app/controllers/todos_controller.rb (hunk 1 above)
Comment (mcgBradley) [reply app/controllers/todos_controller.rb]: This is an error prevention if statement. Without this if statement, if a user was to mark a todo complete or incomplete then the website will crash as without it the code will assume there will be a change in the sort.

File: app/controllers/todos_controller.rb (hunk 2 above)
Comment (mcgBradley) [reply app/controllers/todos_controller.rb]: You are correct, when doing pagination I realised how I can do this as an instance variable and that will be changed. Thank you.

File: app/controllers/todos_controller.rb (hunk 3 above)
Comment (mcgBradley) [reply app/controllers/todos_controller.rb]: One more efficient way I can think of is to implement a hashmap where the keys are the sort_Type and the field inside of the key is the todos.order() argument (such as 'title ASC' or 'created_at DESC').

File: spec/features/user_sorts_todos_by_alphabetical_spec.rb
Comment (mcgBradley) [reply spec/features/user_sorts_todos_by_alphabetical_spec.rb:1]: I agree with this and will refactor these tests to be more optimised. I will also implement edge cases.

File: app/views/todos/index.html.erb
Comment (mcgBradley) [reply app/views/todos/index.html.erb:1]: Will definitely make it look nicer after implementing functionality as I will need to learn how to effectively use CSS but good point.

File: app/controllers/todos_controller.rb (hunk 3 above)
Comment (mcgBradley) [reply app/controllers/todos_controller.rb]: Ok so this will be changed and refactored entirely by just calling and passing directly from the database rather than trying to store data inside of @todos which will make things much smoother.

File: spec/features/user_sorts_todos_spec.rb [hunk 7]
```diff
@@ -0,0 +1,103 @@
+require "rails_helper"
+
+feature ".index" do
+    before do
+        sign_in
```
Comment (v1dya) [inline spec/features/user_sorts_todos_spec.rb:5]: Nice work adding todos in the before block. Since we do a `sign_in` for every test in this file we can have a global `before` block that signs in

File: spec/features/user_sorts_todos_spec.rb [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+        it "no sort" do
+            select_sort("No sort")
+
+            expect("Exercise").to appear_before("Buy milk")
+        end
+    end
+end
+
+feature "edge case" do
```
Comment (v1dya) [inline spec/features/user_sorts_todos_spec.rb]: You are testing the right things, however we should follow best practices so it's easier for other developers to understand how your code works.

We wouldn't call the spec itself as "edge case". We should describe the spec at a high level here and then describe it in detail in the nested blocks.

`context` blocks usually gives more information about the environment in the spec and then `it` block asserts it

An example for showing edge cases in a spec.

```
describe '#destroy' do

  context 'when resource is found' do
    it 'responds with 200' do
      expect_something
    end
    it 'shows the resource' do
      expect_something
    end
  end

  context 'when resource is not found' do
    it 'responds with 404'do
      expect_something
    end
  end

  context 'when resource is not owned' do
    it 'responds with 404' do
      expect_something
    end
  end
end
```

Here the 'when resource is not found' and the 'when resource is not owned' are examples of edge cases when testing the `#destroy` method.

The blocks are meant to read like english. If you see here the `it` blocks are kind of like a sentence and the spec is readable. A sentence could be formed like "When resource is not found it responds with 404 and expects response to equal 404". Every spec does not HAVE to read like english but it's nice to get it close to readable.

File: spec/features/user_sorts_todos_spec.rb [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    before do
+        sign_in
+
+        create_todo "A"
+        create_todo "1"
+        create_todo "@"
+    end
+
+    context "when sorting numbers" do
+        it "with letters" do
```
Comment (v1dya) [inline spec/features/user_sorts_todos_spec.rb]: As an example, this spec could be something like 
```
context "when sorting numbers alphabetically" do
  before do
      select_sort("Sort alphabetically")
  end
  
  it "sorts numbers before letters" do
      expect("1").to appear_before("A")
  end
end
```

File: spec/features/user_visits_homepage_spec.rb (hunk 4 above)
Comment (v1dya) [reply spec/features/user_visits_homepage_spec.rb]: Reminder to not mix changes to unrelated specs into one PR.

File: app/views/todos/index.html.erb
Comment (v1dya) [reply app/views/todos/index.html.erb:1]: Is the CSS going to be in another PR?

File: spec/support/features/todo_helper.rb [hunk 10]
```diff
@@ -5,11 +5,20 @@ def create_todo(todo_title)
         click_on "Submit"
     end
 
+    def display_li()
```
Comment (v1dya) [inline spec/support/features/todo_helper.rb]: We don't need parenthesis at the end of a method call unless you are passing in arguments

**Review by v1dya — COMMENTED** [review]
Great work on refactoring! Had some last few comments on the spec readability. After that I think we can merge ⛴️

File: app/views/todos/index.html.erb
Comment (mcgBradley) [reply app/views/todos/index.html.erb:1]: Yes this CSS will be in a pagination branch PR

File: spec/support/features/todo_helper.rb [hunk 11]
```diff
@@ -21,4 +21,9 @@ def select_sort(sort)
         select sort, from: "sort"
         click_button "Submit"
     end
+
+    def before_sort_select
```
Comment (v1dya) [inline spec/support/features/todo_helper.rb]: [Not Blocking] This method should be named after what the method does. You should be able to know what a method does by reading its name, `before_sort_select` would be hard to understand if you're not in context

File: spec/features/user_sorts_todos_spec.rb [hunk 12]
```diff
@@ -3,98 +3,113 @@
...
         end
     end
-end
 
-feature "edge case" do
-    before do
-        sign_in
-    end
+    context "when sorting no todos alphabetically" do
+        before do
```
Comment (v1dya) [inline spec/features/user_sorts_todos_spec.rb]: Empty before block?

File: spec/features/user_sorts_todos_spec.rb [hunk 13]
```diff
@@ -3,98 +3,113 @@
...
-    before do
-        sign_in
-
-        create_todo ""
-        create_todo "A"
-    end
 
     context "when creating empty todo" do
+        before do
+            create_todo ""
```
Comment (v1dya) [inline spec/features/user_sorts_todos_spec.rb:98]: [Not Blocking] Should you be allowing them to create empty Todos?

File: spec/features/user_visits_homepage_spec.rb (hunk 4 above)
Comment (v1dya) [reply spec/features/user_visits_homepage_spec.rb]: Reminder again to move this to another PR

**Review by v1dya — CHANGES_REQUESTED** [review]
Nice work on the renaming, it's much easier to read now. Couple comments on some lines that don't need to be there and some non blocking comments. Also make sure to use https://www.betterspecs.org/ this as reference for your future specs.

File: spec/features/user_sorts_todos_spec.rb (hunk 13 above)
Comment (mcgBradley) [reply spec/features/user_sorts_todos_spec.rb:98]: You are correct. There should be a validation on the title of these todos. I shall implement this onto an entirely new (but hopefully quick) PR as that is implementation unrelated to sorts.

File: spec/features/user_visits_homepage_spec.rb (hunk 4 above)
Comment (mcgBradley) [reply spec/features/user_visits_homepage_spec.rb]: Yeah I am going to do this now or I will forget... Just have to figure out how to remove this from this PR and move it.

File: spec/support/features/todo_helper.rb (hunk 11 above)
Comment (mcgBradley) [reply spec/support/features/todo_helper.rb]: Ah right I need to be more careful.

File: spec/features/user_sorts_todos_spec.rb (hunk 12 above)
Comment (mcgBradley) [reply spec/features/user_sorts_todos_spec.rb]: Correct )))). Joking, this was a silly mistake I forgot to remove. Thanks for spotting it.

File: spec/features/user_visits_homepage_spec.rb (hunk 4 above)
Comment (mcgBradley) [reply spec/features/user_visits_homepage_spec.rb]: There we go, you will no longer see it )))

**Review by v1dya — APPROVED** [review]
LGTM 🚢
